lib/Sema/CSSimplify.cpp

```diff
@@ -1012,29 +988,34 @@ ConstraintSystem::TypeMatchResult constraints::matchCallArguments(
     const auto &param = params[paramIdx];
     auto paramTy = param.getOldType();
 
-    if (param.isAutoClosure())
-      paramTy = paramTy->castTo<FunctionType>()->getResult();
-
     // Compare each of the bound arguments for this parameter.
     for (auto argIdx : parameterBindings[paramIdx]) {
       auto loc = locator.withPathElement(LocatorPathElt::
                                             getApplyArgToParam(argIdx,
                                                                paramIdx));
       auto argTy = argsWithLabels[argIdx].getOldType();
 
-      // If parameter was marked as `@autoclosure` and argument
-      // is itself `@autoclosure` function type in Swift < 5,
-      // let's fix that up by making it look like argument is
-      // called implicitly.
-      if (param.isAutoClosure() &&
-          isAutoClosureArg(locator.getAnchor(), argIdx)) {
-        argTy = argTy->castTo<FunctionType>()->getResult();
-        cs.increaseScore(SK_FunctionConversion);
-
-        if (cs.getASTContext().isSwiftVersionAtLeast(5)) {
-          auto *fixLoc = cs.getConstraintLocator(loc);
-          if (cs.recordFix(AutoClosureForwarding::create(cs, fixLoc)))
-            return cs.getTypeMatchFailure(loc);
+      bool matchingAutoClosureResult = param.isAutoClosure();
+      if (param.isAutoClosure()) {
+        auto *argExpr = getArgumentExpr(locator.getAnchor(), argIdx);
+        auto *fnType = paramTy->castTo<FunctionType>();
+        if (isAutoClosureArgument(argExpr)) {
+          if (cs.getASTContext().isSwiftVersionAtLeast(5)) {
+            // In Swift >= 5 mode there is no @autoclosure forwarding,
+            // so let's match result types.
+            paramTy = fnType->getResult();
+          } else {
+            // Matching @autoclosure argument to @autoclosure parameter
+            // directly would mean introducting a function conversion
+            // in Swift <= 4 mode.
+            cs.increaseScore(SK_FunctionConversion);
+            matchingAutoClosureResult = false;
+          }
+        } else {
```

Comment:
Can you flatten the 'if' so that the else branch here is take nif `cs.getASTContext().isSwiftVersionAtLeast(5) || !isAutoClosureArgument(argExpr)`

Reply:
Sure, if you think that it would be cleaner. I wanted to keep it this way to make it more explicit that we have different situations and version only affects things when argument is marked as `@autoclosure`.

Reply:
Sure, it's just a tradeoff between minimizing number of cases vs complexity of the boolean expressions. (Also generally if I leave comments on an 'approved' review, they're optional)

Reply:
Thanks for clarification, that's good to know! :)

Reply:
Alright, done!